jdk/test/sun/nio/cs/TestCharsetMapping.java

```diff
@@ -0,0 +1,640 @@
+
+        // mapping entries
+        List<Entry> mappings;
+        Map<Long, Entry> nr;       // bytes -> entry
+        Map<Integer, Entry> c2b;   // cp -> entry
+
+        CharsetInfo(String csName, String clzName) {
+            this.csName = csName;
+            this.clzName = clzName;
+            if (csName.endsWith("_Solaris") ||
+                csName.endsWith("_MS5022X") ||
+                csName.endsWith("_MS932")) {
+                isInternal = true;
+            }
+        }
+
+        private Entry parse(Matcher m) {
+            Entry e = new Entry();
+            e.bb = Long.parseLong(m.group(G_BS), 16);
+            if (e.bb < 0x100)
+                e.bs = new byte[] { (byte)e.bb };
+            else
+                e.bs = parseBytes(m.group(G_BS));
+            e.cp = Integer.parseInt(m.group(G_CP), 16);
+            if (G_CP2 <= m.groupCount() && m.group(G_CP2) != null)
+               e.cp2 = Integer.parseInt(m.group(G_CP2), 16);
+            else
+               e.cp2 = 0;
+            return e;
+        }
+
+        boolean loadMappings(Path dir) throws IOException {
+            // xxx.map
+            Path path = dir.resolve(clzName + ".map");
+            if (!Files.exists(path)) {
+                return false;
+            }
+            Matcher m = ptn.matcher("");
+            mappings = Files.lines(path)
+                .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                .map(ln -> parse(m))
+                .filter(e -> e.cp != UNMAPPABLE)  // non-mapping
+                .collect(Collectors.toList());
+            // xxx.nr
+            path = dir.resolve(clzName + ".nr");
+            if (Files.exists(path)) {
+                nr = Files.lines(path)
+                    .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                    .map(ln -> parse(m))
+                    .collect(Collectors.toMap(e -> e.bb, Function.identity()));
+            }
+            // xxx.c2b
+            path = dir.resolve(clzName + ".c2b");
+            if (Files.exists(path)) {
+                c2b = Files.lines(path)
+                    .filter(ln -> !ln.startsWith("#") && m.reset(ln).lookingAt())
+                    .map(ln -> parse(m))
+                    .collect(Collectors.toMap(e -> e.cp, Function.identity()));
+            }
+            return true;
+        }
+    }
+
+    private static Set<CharsetInfo> charsets(Path cslist, boolean sbcs) throws IOException {
+        Set<CharsetInfo> charsets = new LinkedHashSet<>();
+        Iterator<String> itr = Files.readAllLines(cslist).iterator();
+
+        while (itr.hasNext()) {
+            String line = itr.next();
+            if (line.startsWith("#") || line.length() == 0) {
+                continue;
+            }
+            String[] tokens = line.split("\\s+");
+            if (tokens.length < 5) {
+                continue;
+            }
+            CharsetInfo cs = new CharsetInfo(tokens[1], tokens[0]);
```

Comment:
Small nit, could you add comments to the constructor args like this:
```
new CharsetInfo(/*csName*/ tokens[1], /*clzName*/ tokens[0]);
```

(We are different from the JDK 10 version anyway in this hunk)

Reply:
Done, though it should also now be clearer from having the format in there.